bin/monitor.ml

```diff
@@ -0,0 +1,236 @@
+  let progress p = Progress p
+end
+
+module State = struct
+  (** Internal state of the render loop. *)
+
+  type t =
+    { mutable quit : bool
+    ; mutable diagnostics : Diagnostic.t Diagnostic.Id.Map.t
+    ; mutable jobs : Job.t Job.Id.Map.t
+    ; mutable progress : Progress.t
+    }
+
+  let init =
+    { quit = false
+    ; diagnostics = Diagnostic.Id.Map.empty
+    ; jobs = Job.Id.Map.empty
+    ; progress = Waiting
+    }
+  ;;
+
+  let done_status ~complete ~remaining ~failed ~jobs =
+    Pp.textf
+      "Done: %d%% (%d/%d, %d left%s) (jobs: %d)"
+      (if complete + remaining = 0 then 0 else complete * 100 / (complete + remaining))
+      complete
+      (complete + remaining)
+      remaining
+      (match failed with
+       | 0 -> ""
+       | failed -> sprintf ", %d failed" failed)
+      jobs
+  ;;
+
+  let waiting_for_file_system_changes message =
+    Pp.seq message (Pp.verbatim ", waiting for filesystem changes...")
+  ;;
+
+  let restarting_current_build message =
+    Pp.seq message (Pp.verbatim ", restarting current build...")
+  ;;
+
+  let had_errors number =
+    match number with
+    | 1 -> Pp.verbatim "Had 1 error"
+    | n -> Pp.textf "Had %d errors" n
+  ;;
+
+  let status (state : t) =
+    Console.Status_line.set
+      (Live
+         (fun () ->
+           match (state.progress : Progress.t) with
+           | Waiting -> Pp.verbatim "Initializing..."
+           | In_progress { complete; remaining; failed } ->
+             done_status
+               ~complete
+               ~remaining
+               ~failed
+               ~jobs:(Job.Id.Map.cardinal state.jobs)
+           | Interrupted ->
+             Pp.tag User_message.Style.Error (Pp.verbatim "Source files changed")
+             |> restarting_current_build
+           | Success ->
+             Pp.tag User_message.Style.Success (Pp.verbatim "Success")
+             |> waiting_for_file_system_changes
+           | Failed ->
+             let number = Diagnostic.Id.Map.cardinal state.diagnostics in
+             Pp.tag User_message.Style.Error (had_errors number)
+             |> waiting_for_file_system_changes))
+  ;;
+
+  let render_state (state : t) =
+    Diagnostic.Id.Map.iter state.diagnostics ~f:(fun d ->
+      Console.print_user_message (Diagnostic.to_user_message d));
+    status state
+  ;;
+
+  let maybe_reset_diagnostics (state : t) =
+    match (state.progress : Progress.t) with
+    | Interrupted ->
+      state.diagnostics <- init.diagnostics;
+      Console.reset ();
+      render_state state
+    | _ -> status state
+  ;;
+end
+
+(** A generic loop that fetches events from a [poller] and writes them to the
+    [event] mvar. *)
+let fetch_loop (event : Event.t Fiber.Mvar.t) (poller : 'a Client.Stream.t) ~f =
+  let rec loop () =
+    let* next = Client.Stream.next poller in
+    match next with
+    | Some payload -> Fiber.Mvar.write event (f payload) >>> loop ()
+    | None -> Fiber.Mvar.write event Event.Poller_disconnected
```

Comment:
This should also set the quit flag. We should signal disconnects as soon as we can